packages/ui/src/components/MessageInput/InputToolbar.tsx

```diff
@@ -0,0 +1,117 @@
+import {
+  EditorBridge,
+  useBridgeState,
+  useKeyboard,
+} from '@10play/tentap-editor';
+import React from 'react';
+import { FlatList, Platform, TouchableOpacity } from 'react-native';
+
+import { Image, View } from '../../core';
+import { EditLinkBar } from './EditLinkBar';
+import {
+  DEFAULT_TOOLBAR_ITEMS,
+  HEADING_ITEMS,
+  ToolbarContext,
+  type ToolbarItem,
+} from './toolbarActions';
+
+interface ToolbarProps {
+  editor: EditorBridge;
+  hidden?: boolean;
+  items?: ToolbarItem[];
+}
+
+export function InputToolbar({
+  editor,
+  hidden = undefined,
+  items = DEFAULT_TOOLBAR_ITEMS,
+}: ToolbarProps) {
+  const editorState = useBridgeState(editor);
+  const { isKeyboardUp } = useKeyboard();
+  const [toolbarContext, setToolbarContext] = React.useState<ToolbarContext>(
+    ToolbarContext.Main
+  );
+
+  const hideToolbar =
+    hidden === undefined ? !isKeyboardUp || !editorState.isFocused : hidden;
+
+  const args = {
+    editor,
+    editorState,
+    setToolbarContext,
+    toolbarContext,
+  };
+
+  switch (toolbarContext) {
+    case ToolbarContext.Main:
+    case ToolbarContext.Heading:
+      return (
+        <FlatList
```

Comment:
This list could use some optimization -- it's rerendering frequently during editing. I'd memoize everything you can, maybe break out the item component. `FlatList` might also be overkill for this use case.

<img width="912" alt="Screenshot 2024-05-30 at 5 12 54 PM" src="https://github.com/tloncorp/tlon-apps/assets/357216/d0b1db0e-2573-4772-bb70-c4776068c56b">

Reply:
Memoized a bunch of stuff and added some perf related changes for FlatList, got it down to between 5 and 8ms per character. Unfortunately we do need to update it on each change from the editor in order to show the current enabled/disabled state for the buttons.